From bd08c273dd67939cbce1a21fcb57e7a42420fa18 Mon Sep 17 00:00:00 2001
From: Roman Meshkevich <roman.meshkevich@cogentembedded.com>
Date: Wed, 14 Sep 2016 20:22:09 +0300
Subject: [PATCH 2/5] v4l2-renderer: clear dmabuf fd when a kms buffer is
 destoryed.

Clear a dmabuf fd stored in the surface state when a kms buffer is
destroyed. This prevents from referring to invalidated dmabuf fd when
the same fd is reused for the different purpose.
---
 src/v4l2-renderer-device.h |  2 +-
 src/v4l2-renderer.c        | 27 +++++++++++++++++++++++++++
 2 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/src/v4l2-renderer-device.h b/src/v4l2-renderer-device.h
index c2c8c27..b615142 100644
--- a/src/v4l2-renderer-device.h
+++ b/src/v4l2-renderer-device.h
@@ -87,7 +87,7 @@ struct v4l2_surface_state {
 	struct wl_listener buffer_destroy_listener;
 	struct wl_listener surface_destroy_listener;
 	struct wl_listener renderer_destroy_listener;
-
+	struct wl_listener kms_buffer_destroy_listener;
 #ifdef V4L2_GL_FALLBACK
 	void *gl_renderer_state;
 
diff --git a/src/v4l2-renderer.c b/src/v4l2-renderer.c
index e4a57c1..db79b40 100644
--- a/src/v4l2-renderer.c
+++ b/src/v4l2-renderer.c
@@ -1023,6 +1023,17 @@ error:
 	return -1;
 }
 
+static void
+kms_buffer_state_handle_buffer_destroy(struct wl_listener *listener, void *data)
+{
+	struct v4l2_surface_state *vs;
+
+	vs = container_of(listener, struct v4l2_surface_state,
+			  kms_buffer_destroy_listener);
+	vs->planes[0].dmafd = 0;
+	vs->kms_buffer_destroy_listener.notify = NULL;
+}
+
 static int
 v4l2_renderer_attach_dmabuf(struct v4l2_surface_state *vs, struct weston_buffer *buffer)
 {
@@ -1133,6 +1144,15 @@ v4l2_renderer_attach_dmabuf(struct v4l2_surface_state *vs, struct weston_buffer
 	if (device_interface->attach_buffer(vs) == -1)
 		return -1;
 
+	if (vs->kms_buffer_destroy_listener.notify) {
+		wl_list_remove(&vs->kms_buffer_destroy_listener.link);
+		vs->kms_buffer_destroy_listener.notify = NULL;
+	}
+	vs->kms_buffer_destroy_listener.notify
+		= kms_buffer_state_handle_buffer_destroy;
+	wl_resource_add_destroy_listener(kbuf->resource,
+					 &vs->kms_buffer_destroy_listener);
+
 	DBG("%s: %dx%d buffer attached (dmabuf=%d, stride=%d).\n", __func__, kbuf->width, kbuf->height, kbuf->fd, kbuf->stride);
 
 	return 0;
@@ -1194,6 +1214,13 @@ v4l2_renderer_surface_state_destroy(struct v4l2_surface_state *vs)
 		vs->buffer_destroy_listener.notify = NULL;
 	}
 
+	vs->surface->renderer_state = NULL;
+
+	if (vs->kms_buffer_destroy_listener.notify) {
+		wl_list_remove(&vs->kms_buffer_destroy_listener.link);
+		vs->kms_buffer_destroy_listener.notify = NULL;
+	}
+
 	// TODO: Release any resources associated to the surface here.
 
 	weston_buffer_reference(&vs->buffer_ref, NULL);
-- 
2.7.4

